File: 프로그래머스/3/12927. 야근 지수/야근 지수.cpp
#include <string>
#include <vector>
#include <queue>
using namespace std;
long long solution(int n, vector<int> works) {
    long long answer = 0;
    priority_queue<int> pq;
    for (int i = 0; i < works.size(); i++) pq.push(works[i]);
    while (n != 0) {
        n--;
        int top = pq.top(); pq.pop();
        if (top == 0) return 0;
        pq.push(--top);

    }
    while (!pq.empty()) {
        int top = pq.top(); pq.pop();
        answer += (top * top);
    }
    return answer;
}

File: 프로그래머스/3/12927. 야근 지수/야근 지수.py
def solution(n, works):
    from collections import Counter
    work_remain_list = sorted(Counter(works).items())
    work_remain_list = [list(ele) for ele in work_remain_list]

    while n > 0 :
        time, amount = work_remain_list.pop()
        if n >= amount :
            if work_remain_list and work_remain_list[-1][0] == time -1:
                work_remain_list[-1][1] += amount
            else:
                work_remain_list.append([time-1,amount])
            n -= amount
        else:
            if work_remain_list and work_remain_list[-1][0] == time -1:
                work_remain_list[-1][1] += n
            else:
                work_remain_list.append([time-1,n])
            work_remain_list.append([time,amount-n])
            n -= amount

    if work_remain_list[-1][0] <=0: return 0
    result = 0
    for time, amount in work_remain_list:
        result+= amount*time**2
    return result